crates/c-api/CMakeLists.txt

```
@@ -1,140 +1,174 @@
+option(WASMTIME_USER_CARGO_BUILD_OPTIONS "Additional cargo flags (such as --features) to apply to the build command" OFF)
 option(BUILD_SHARED_LIBS "Build using shared libraries" OFF)
 option(WASMTIME_ALWAYS_BUILD "If cmake should always invoke cargo to build wasmtime" ON)
 option(WASMTIME_FASTEST_RUNTIME "Set flags designed to optimize runtime performance" OFF)
+option(WASMTIME_TARGET "Rust target to build for" OFF)
 
-if(WASMTIME_FASTEST_RUNTIME)
-	set(WASMTIME_BUILD_TYPE_FLAG "--profile=fastest-runtime")
-	set(WASMTIME_BUILD_TYPE "fastest-runtime")
-	set(CARGO_PROFILE_PANIC CARGO_PROFILE_RELEASE_PANIC)
-else()
-	if(CMAKE_BUILD_TYPE STREQUAL "Release" OR
-	    CMAKE_BUILD_TYPE STREQUAL "MinSizeRel" OR
-	    CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
-		set(WASMTIME_BUILD_TYPE_FLAG "--release")
-		set(WASMTIME_BUILD_TYPE "release")
-		set(CARGO_PROFILE_PANIC CARGO_PROFILE_RELEASE_PANIC)
-	else()
-		set(WASMTIME_BUILD_TYPE "debug")
-		set(CARGO_PROFILE_PANIC CARGO_PROFILE_DEBUG_PANIC)
-    endif()
+if(NOT WASMTIME_TARGET)
+  execute_process(COMMAND rustc -vV OUTPUT_VARIABLE RUSTC_VERSION)
+  string(REGEX MATCH "host: ([^ \n]*)" RUSTC_HOST ${RUSTC_VERSION})
+  string(STRIP ${CMAKE_MATCH_1} WASMTIME_TARGET)
 endif()
 
-if(ANDROID)
-	# TODO wasmtime only supports arm64-v8a right now
-	if(ANDROID_ABI STREQUAL "armeabi-v7a")
-		set(ANDROID_TARGET "armv7-linux-androideabi")
-		set(ANDROID_ARCH_SHORT "arm")
-	elseif(ANDROID_ABI STREQUAL "arm64-v8a")
-		set(ANDROID_TARGET "aarch64-linux-android")
-		set(ANDROID_ARCH_SHORT "aarch64")
-	elseif(ANDROID_ABI STREQUAL "x86")
-		set(ANDROID_TARGET "i686-linux-android")
-		set(ANDROID_ARCH_SHORT "i386")
-	elseif(ANDROID_ABI STREQUAL "x86_64")
-		set(ANDROID_TARGET "x86_64-linux-android")
-		set(ANDROID_ARCH_SHORT "x86_64")
-	endif()
-
-	set(WASMTIME_BUILD_TARGET "--target=${ANDROID_TARGET}")
+if(NOT WASMTIME_USER_CARGO_BUILD_OPTIONS)
+  set(WASMTIME_USER_CARGO_BUILD_OPTIONS "")
 endif()
 
-if (BUILD_SHARED_LIBS AND ANDROID)
-	message(FATAL_ERROR "Wasmtime cannot be built with BUILD_SHARED_LIBS on Android")
-endif()
+set(WASMTIME_FEATURES "--no-default-features")
 
-if(BUILD_SHARED_LIBS)
-	if(WIN32)
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/wasmtime.dll.lib)
-	elseif(APPLE)
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/libwasmtime.dylib)
-	else()
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/libwasmtime.so)
-	endif()
+option(WASMTIME_DISABLE_ALL_FEATURES
+       "disable all features by default instead of enabling them"
+       OFF)
+
+macro(feature cmake_name rust_name default)
+  if(${default})
+    if(${WASMTIME_DISABLE_ALL_FEATURES})
+      set(feature_default OFF)
+    else()
+      set(feature_default ON)
+    endif()
+  else()
+    set(feature_default OFF)
+  endif()
```

Nitpick: arguably cleaner

```suggestion
  set(feature_default "${default} AND NOT ${WASMTIME_DISABLE_ALL_FEATURES}")
```

---

I don't think I fully understand cmake evaluation contexts and strings and what's a variable when, but when I use your suggestion it looks like it sets `feature_default` to the string `ON AND NOT OFF` and I'm not sure why but the default build has all features off.

I'm also not sure if I'm supposed to do something like `if(${${cmake_name}})` below instead of just `if(${cmake_name})` (sorry I'm very new to cmake)

---

Actually I retried it and my suggestion was wrong. My bad, I should've investigated better beforehand. Ignore my comment, your method is the way to go.